presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/JoinReorderingOptimizer.java

```diff
@@ -0,0 +1,119 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.statistics.PlanNodeStatistics;
+import com.facebook.presto.statistics.StatisticsCalculator;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class JoinReorderingOptimizer
+        implements PlanOptimizer
+{
+    private final StatisticsCalculator statisticsCalculator;
+
+    public JoinReorderingOptimizer(StatisticsCalculator statisticsCalculator)
+    {
+        this.statisticsCalculator = requireNonNull(statisticsCalculator, "statisticsCalculator can not be null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(session), plan);
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+
+        public Rewriter(Session session)
+        {
+            this.session = session;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode leftRewritten = context.defaultRewrite(node.getLeft());
+            PlanNode rightRewritten = context.defaultRewrite(node.getRight());
+
+            PlanNodeStatistics leftStatistics = statisticsCalculator.calculateStatisticsForNode(session, leftRewritten);
+            PlanNodeStatistics rightStatistics = statisticsCalculator.calculateStatisticsForNode(session, rightRewritten);
+
+            if (!leftStatistics.getOutputRowsCount().isValueUnknown()
+                    && !rightStatistics.getOutputRowsCount().isValueUnknown()
+                    && (leftStatistics.getOutputRowsCount().getValue() < rightStatistics.getOutputRowsCount().getValue())) {
```

Comment:
compare `datasize` instead of rows count. If there is no data size, then fallback to rows count.